asc2log.c

```diff
@@ -71,7 +71,9 @@ void print_usage(char *prg)
 	fprintf(stderr, "\t-O <outfile>\t(default stdout)\n");
 }
 
-void prframe(FILE *file, struct timeval *tv, int dev, struct canfd_frame *cf, unsigned int max_dlen, char *extra_info) {
+void prframe(FILE *file, struct timeval *tv, int dev, struct canfd_frame *cf, char *extra_info) {
+
+	static char abuf[BUFLEN];
```

Comment:
The idea was that the quite big data structures like the afrbuf[AFRSZ]  (6300 bytes) and the cu_t cu (2060 bytes) should not be allocated on the stack or the heap (potentially every time the function is executed).

Therefore I entirely removed any local (non-static) ASCII buffers from the lib.c code and let the single threaded(!) applications provide the buffers. This removed fprint_(long_)canframe() in favor of sprint_(long_)canframe().

Do you still think allocating static buffers for the ASCII representation and the cu_t union is a bad design choice then?

Reply:
@marckleinebudde I added 4 patches three days ago to follow your review recommendations. But not this "static buffer" topic. Are you fine with my justification above?

Reply:
Ok, `asc22log` is the application not the library. I probably mixed that up, as the function `prframe()` is not declared `static`, which suggests that it's used outside of this compilation unit.

> Therefore I entirely removed any local (non-static) ASCII buffers from the `lib.c` code and let the single threaded(!) applications provide the buffers. This removed `fprint_(long_)canframe()` in favor of `sprint_(long_)canframe()`.

This is a good idea, please let the caller provide buffer _and_ length of buffer, i.e. make it `snprintf_*()` functions. In the caller however, I'd still not use `static` buffers in functions. If you call the function the first time, and the stack has never grown this big, real memory will be provided via pagefaults. Once the stack is backed by real memory there's no runtime overhead. For a more detailed picture we can talk about `mlockall()`, memory over commitment, fixed stack size for threads, swap, caching, etc...

Reply:
> Ok, `asc22log` is the application not the library. I probably mixed that up, as the function `prframe()` is not declared `static`, which suggests that it's used outside of this compilation unit.

Would you generally suggest to make all local functions inside applications 'static' ? 
If so this could be a separate change on the can-utils repo later ...

> > Therefore I entirely removed any local (non-static) ASCII buffers from the `lib.c` code and let the single threaded(!) applications provide the buffers. This removed `fprint_(long_)canframe()` in favor of `sprint_(long_)canframe()`.
> 
> This is a good idea, please let the caller provide buffer _and_ length of buffer, i.e. make it `snprintf_*()` functions. 

That's a good idea too. It creates confidence in sprint_(long_)canframe() and the caller knows best which buffer is provided anyway.

> In the caller however, I'd still not use `static` buffers in functions. If you call the function the first time, and the stack has never grown this big, real memory will be provided via pagefaults. Once the stack is backed by real memory there's no runtime overhead. For a more detailed picture we can talk about `mlockall()`, memory over commitment, fixed stack size for threads, swap, caching, etc...

But in the end a pre-defined static buffer reduces all these types of allocation pagefault complexity (when a static buffer is possible). So I'm still not really convinced ;-D

Reply:
> Would you generally suggest to make all local functions inside applications 'static' ?
If so this could be a separate change on the can-utils repo later ...

ACK

> That's a good idea too. It creates confidence in sprint_(long_)canframe() and the caller knows best which buffer is provided anyway.

`s`**`n`**`printf_` with emphasis on the `n`

> But in the end a pre-defined static buffer reduces all these types of allocation pagefault complexity (when a static buffer is possible). So I'm still not really convinced ;-D

A static buffer lives in the BSS and this is faulted into during the initialization of the C runtime, at it's `memset()`ed to `0x0`.